cmake/rocprim-backward-compat.cmake

```diff
@@ -0,0 +1,309 @@
+set(PROJECT_WRAPPER_DETAIL_RELPATH      "${include_statements}#include \"../../../../include/rocprim/detail")
+set(PROJECT_WRAPPER_DEVICE_RELPATH      "${include_statements}#include \"../../../../include/rocprim/device")
+set(PROJECT_WRAPPER_INTRINSICS_RELPATH  "${include_statements}#include \"../../../../include/rocprim/intrinsics")
+set(PROJECT_WRAPPER_ITERATOR_RELPATH    "${include_statements}#include \"../../../../include/rocprim/iterator")
+set(PROJECT_WRAPPER_TYPES_RELPATH       "${include_statements}#include \"../../../../include/rocprim/types")
+set(PROJECT_WRAPPER_THREAD_RELPATH      "${include_statements}#include \"../../../../include/rocprim/thread")
+set(PROJECT_WRAPPER_WARP_RELPATH        "${include_statements}#include \"../../../../include/rocprim/warp")
+set(PROJECT_WRAPPER_BLK_DET_RELPATH     "${include_statements}#include \"../../../../../include/rocprim/block/detail")
+set(PROJECT_WRAPPER_DEV_DET_RELPATH     "${include_statements}#include \"../../../../../include/rocprim/device/detail")
+set(PROJECT_WRAPPER_DEV_SPE_RELPATH     "${include_statements}#include \"../../../../../include/rocprim/device/specialization")
+set(PROJECT_WRAPPER_ITE_DET_RELPATH     "${include_statements}#include \"../../../../../include/rocprim/iterator/detail")
+set(PROJECT_WRAPPER_WAR_DET_RELPATH     "${include_statements}#include \"../../../../../include/rocprim/warp/detail")
+
+#Include Guard for wrapper (include) files generated
+set(PROJECT_WRAPPER_MAIN_INC_GUARD      "${include_guard}ROCPRIM_WRAPPER_MAIN_INC")
+set(PROJECT_WRAPPER_BLOCK_GUARD     	"${include_guard}ROCPRIM_WRAPPER_BLOCK_INC")
+set(PROJECT_WRAPPER_DETAIL_GUARD        "${include_guard}ROCPRIM_WRAPPER_DETAIL_INC")
+set(PROJECT_WRAPPER_DEVICE_GUARD        "${include_guard}ROCPRIM_WRAPPER_DEVICE_INC")
+set(PROJECT_WRAPPER_INTRINSICS_GUARD    "${include_guard}ROCPRIM_WRAPPER_INTRINSICS_INC")
+set(PROJECT_WRAPPER_ITERATOR_GUARD      "${include_guard}ROCPRIM_WRAPPER_ITERATOR_INC")
+set(PROJECT_WRAPPER_TYPES_GUARD         "${include_guard}ROCPRIM_WRAPPER_TYPES_INC")
+set(PROJECT_WRAPPER_THREAD_GUARD        "${include_guard}ROCPRIM_WRAPPER_THREAD_INC")
+set(PROJECT_WRAPPER_WARP_GUARD          "${include_guard}ROCPRIM_WRAPPER_WARP_INC")
+set(PROJECT_WRAPPER_BLK_DET_GUARD       "${include_guard}ROCPRIM_WRAPPER_BLK_DET_INC")
+set(PROJECT_WRAPPER_DEV_DET_GUARD       "${include_guard}ROCPRIM_WRAPPER_DEV_DET_INC")
+set(PROJECT_WRAPPER_DEV_SPE_GUARD       "${include_guard}ROCPRIM_WRAPPER_DEV_SPE_INC")
+set(PROJECT_WRAPPER_ITE_DET_GUARD       "${include_guard}ROCPRIM_WRAPPER_ITE_DET_INC")
+set(PROJECT_WRAPPER_WAR_DET_GUARD       "${include_guard}ROCPRIM_WRAPPER_WAR_DET_INC")
+
+
+
+# Function for Generating Wrapper Headers for Backward Compatibilty
+# Generates wrapper for all *.h files in include/rocprim folder
+# No Arguments to function.
+# Wrapper are generated under rocprim/include/rocprim.
+function (package_gen_bkwdcomp_hdrs)
```

Comment:
I think this function is massively overcomplicated, and could very easily be replaced with something simpler. For example,
```cmake
function(package_header_wrappers) # CMake function names usually avoid abbreviations
    file(GLOB_RECURSE include_files RELATIVE "${PROJECT_FREORG_MAIN_INC_DIR}" *.hpp)
    foreach(include_file ${include_files})
        get_filename_component(file_name ${include_file} NAME_WE)
        get_filename_component(file_path ${include_file} DIRECTORY)
        string(REPLACE '/' ';' path_dirs "${file_path}")
        set(rel_path "../../../include/rocprim")
        set(guard "ROCPRIM_WRAPPER")
        foreach(subdir IN_LIST path_dirs)
            if(NOT (subdir STREQUAL '' OR subdir STREQUAL '.'))
                set(rel_path "../${rel_path}/${subdir}")
                string(TOUPPER ${subdir} subdir_uc)
                set(guard "${guard}_${subdir_uc})
            endif()
        endforeach()
        string(TOUPPER ${file_name} file_name_uc)
        set(include_guard "${guard}_${file_name_uc}_HPP")
        set(include_statements "${rel_path}/${file_name}.hpp")
        configure_file(${PROJECT_WRAPPER_TEMPLATE_FILE} "${PROJECT_BINARY_DIR}/include/rocprim")
    endforeach()
endfunction()
```

Reply:
Thanks for the comments/ref code shared.  updated on the latest version